### DISCUSSION TITLE
When updating coreshop from version 2.2 to version 3

### DISCUSSION THREAD

**Discussion by claudesimo1990** [pimcore/pimcore#15510] [General]
The classes table has been modified. The Products no longer have the reference number 17 but rather cs_product so new tables are created with the extension _cs_product_ instead of _17_. with this change my data (Products) is no longer displayed on the site. because the requests are made on the new table with the extension (cs_product). is there a solution? i would like not to do this manually, as it involves a lot of sensitive data.

**Comment by dpfaffenbauer** [comment]
@claudesimo1990 please check your original class definitions id values and compare them to what they are after the update. if they changed, please revert the line to what it was before.

**Comment by claudesimo1990** [comment]
@dpfaffenbauer

After Coreshop Update
the definition_CoreShopProduct.php looks like:
`protected $o_classId = "cs_product";
protected $o_className = "CoreShopProduct";`

Before Update
`protected $o_classId = "17";
protected $o_className = "CoreShopProduct";`

![Clipboard - 11  Juli 2023 12_55](https://github.com/pimcore/pimcore/assets/45970298/b800d37e-e77f-4698-9cdd-fb44840189df)
![Clipboard - 11  Juli 2023 12_58](https://github.com/pimcore/pimcore/assets/45970298/cd7e69af-48e9-42b2-b6fe-86f7e71d4591)

After the update, we will receive a new empty table for the products with the classID cs_product, in addition to the old table. However, our products will still be located in the old table (object_localized_query_17_de_DE).

**Reply by dpfaffenbauer** [reply]
@claudesimo1990 not sure how you did update the class-definitions, but change the id back to 17 (in the file var/classes/CoreShopProduct_definition.php).

Please also check all other CoreShop Definition files.

Once you checked them, run ´bin/console pimcore:build:classes` and you should have your data back

**Comment by claudesimo1990** [comment]
@dpfaffenbauer thank you very much now it works :)

But we must update the class-definitions so that the migrations could pass through or is there another solution for the migration (in the vendor)?

We also have the problem that the controllers in the document_page table are not adjusted during the migrations. In the new Coreshop version (3), the controllers only work if there is no '@' in the DB.  

![Clipboard - 12  Juli 2023 08_14](https://github.com/pimcore/pimcore/assets/45970298/8dbf92a8-0495-498a-8c98-c91c8cd4ccaa)

**Reply by dpfaffenbauer** [reply]
> But we must update the class-definitions so that the migrations could pass through or is there another solution for the migration (in the vendor)?

I did several updates of CoreShop and I never had the Problem that the CoreShop Migrations changed the class-ids....

> We also have the problem that the controllers in the document_page table are not adjusted during the migrations. In the new Coreshop version (3), the controllers only work if there is no '@' in the DB.

you have to run `bin/console migration:controller-reference` on Pimcore 6.9 before updating to Pimcore X. The Command does NOT exist in Pimcore X, so you have to do it on Pimcore 6.9.

**Reply by claudesimo1990** [reply]
@dpfaffenbauer 

> you have to run bin/console migration:controller-reference on Pimcore 6.9 before updating to Pimcore X. The Command does NOT exist in Pimcore X, so you have to do it on Pimcore 6.9. 

Thanks for the hint :)

> I did several updates of CoreShop and I never had the Problem that the CoreShop Migrations changed the class-ids....

if I run bin/console `doctrine:migrations:migrate` Command I get this problem. They have problem with the class-id in the class-definitions 

![Clipboard - 12  Juli 2023 10_10](https://github.com/pimcore/pimcore/assets/45970298/76fe115f-649c-4942-a3d7-71e167f39e7b)

**Reply by dpfaffenbauer** [reply]
@claudesimo1990 ok, now it is getting to specific, I don't know what you did before and how the setup and what has been adapted.

**Comment by M4X1L** [comment]
Hi @dpfaffenbauer ,
I am currently facing an issue regarding the management of different themes for Coreshop. I have multiple shops and depending on which store they should render different templates.

In Coreshop 2, I have solved this by creating a separate theme directory for each shop, which contains the overridden templates. This allows me to customize the templates for each individual shop.

config.yaml
```
twig:
    paths:
        '%kernel.project_dir%/templates/themes/common': 'CORESHOP_COMMON'
```

services.yaml
```
    app.coreshop.store.context.request.document_based:
        class: AppBundle\CoreShop\Store\Context\DocumentBasedRequestRequestResolver
        arguments:
            - '@coreshop.repository.store'
        tags:
            - { name: coreshop.context.store.request_based.resolver, priority: 300 }
```

How does it work in Coreshop 3?

**Reply by dpfaffenbauer** [reply]
basically the same... CoreShop 3 uses Sylius Theme Bundle for theming. You can use theme inheritance to have a common theme.

**Comment by claudesimo1990** [comment]
@dpfaffenbauer 

Good day, I have another problem with rebuilding the classes. when I run the command `bin/console pimcore:build:classes` and `bin/console pimcore:deployment:classes-rebuild -c -n` I don't get any errors displayed but on the page I see the error with the return type. is there another command to rebuild the classes with the return type of the individual methods in the classes?

![image](https://github.com/pimcore/pimcore/assets/45970298/31979b01-6214-4bf7-99ed-95397fdab81b)

**Reply by dpfaffenbauer** [reply]
CoreShop added return-types with CoreShop 3, please enable that in your class-definitions by setting 'generateTypeDeclarations' to true in all CoreShop definition files in `var/classes/definition_CoreShop*.php`

**Comment by claudesimo1990** [comment]
@dpfaffenbauer 

I found this migration guide online, please I want to know how can I do this without adjusting the vendor directories ?

![image](https://github.com/pimcore/pimcore/assets/45970298/b63d4b19-df02-4a39-8089-e2f144807977)

https://github.com/coreshop/CoreShop/issues/2161
